Improve score by supporting `extra_phrase` for extra words in rules

alok1304 opened the pull request:
Add new phrases like `extra_phrase` this is special for `extra-words`. This phrase is represented in the format `[[n]]`, where n indicates the maximum number of `extra-words` allowed at that position in the rule.

If `extra-words` appear at the correct position and their count does not exceed the allowed limit `n`, then the score is increased to `100`.

<!-- Delete Template sections if unneccesary -->
<!-- Add issue number here (We encourage you to create the Issue First) -->
<!-- You can also link the issue in Commit Messages -->

Reference #4420 

<!--
Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!
* [x] - Checked Box
* [ ] - Unchecked Box
-->

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [ ] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁
* [ ] Updated documentation pages (if applicable)
* [ ] Updated CHANGELOG.rst (if applicable)
<!--
We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged**
if your tests fail at first!
If tests do fail, click on the red `X` to learn why by reading the logs.
Thanks!
-->

<!-- Don't forget to Signoff -->
Signed-off-by: Alok Kumar [alokkumarjipura9973@gmail.com](mailto:alokkumarjipura9973@gmail.com)

AyanSinhaMahapatra committed:
Resolve license references to files beside package manifest

Reference: https://github.com/aboutcode-org/scancode-toolkit/issues/3707
Reference: https://github.com/aboutcode-org/scancode-toolkit/pull/3970
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>
Signed-off-by: Alok Kumar <alokkumarjipura9973@gmail.com>

leoreinmann committed:
fix: change version number in field "name" for license "LiLiQ-R-1.1",

Signed-off-by: Leo Reinmann <leo.reinmann@siemens.com>
Signed-off-by: Alok Kumar <alokkumarjipura9973@gmail.com>

AyanSinhaMahapatra committed:
Skip private packages in `--todo` detection issues

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>
Signed-off-by: Alok Kumar <alokkumarjipura9973@gmail.com>

AyanSinhaMahapatra committed:
Fix windows package license detection

* Do not include copyright statement in license detection
* Handle cases of license/trademarks not being present correctly

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>
Signed-off-by: Alok Kumar <alokkumarjipura9973@gmail.com>

AyanSinhaMahapatra committed:
Show matched test diagnostics for `--todo` issues

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>
Signed-off-by: Alok Kumar <alokkumarjipura9973@gmail.com>

AyanSinhaMahapatra committed:
Make `--todo` output more prominent in results

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>
Signed-off-by: Alok Kumar <alokkumarjipura9973@gmail.com>

AyanSinhaMahapatra committed:
Detect community files to ignore license detections

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>
Signed-off-by: Alok Kumar <alokkumarjipura9973@gmail.com>

AyanSinhaMahapatra committed:
Add misc license detection improvements

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>
Signed-off-by: Alok Kumar <alokkumarjipura9973@gmail.com>

alok1304 committed:
Add new phrase for 'extra-words` in rules

Add new phrases like `extra_phrase` this is special for extra-words.
This phrase is represented in the format [[n]], where n indicates the maximum number of extra-words allowed at that position in the rule.

If extra-words appear at the correct position and their count does not exceed the allowed limit `n`, then the score is increased to `100`.

Signed-off-by: Alok Kumar <alokkumarjipura9973@gmail.com>

leoreinmann committed:
Fix "key" for code-credit-license-1.0-0 (#4416)

* fix: change version number in field "key" for code-credit-license-1.0-0
* fix: change file name for license "code-credit-license-1.0.0"

Signed-off-by: Leo Reinmann <leo.reinmann@siemens.com>
Signed-off-by: Alok Kumar <alokkumarjipura9973@gmail.com>

AyanSinhaMahapatra committed:
Return `sha1_git` checksum for all files as info

Reference: https://github.com/aboutcode-org/scancode-toolkit/issues/624
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>
Signed-off-by: Alok Kumar <alokkumarjipura9973@gmail.com>

AyanSinhaMahapatra committed:
Use new commoncode v32.3.0

* Fix click compatibility issues

Reference: https://github.com/aboutcode-org/scancode-toolkit/issues/4369
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>
Signed-off-by: Alok Kumar <alokkumarjipura9973@gmail.com>

AyanSinhaMahapatra committed:
Address review comments and add misc refactors

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>
Signed-off-by: Alok Kumar <alokkumarjipura9973@gmail.com>

AyanSinhaMahapatra commented on tests/licensedcode/data/plugin_license/extra-words/scan-extra-words-2-aho-license.expected.json:
```diff
@@ -56,10 +54,8 @@
-            "extra-words"
-          ],
-          "identifier": "bsd_new-fbfc5955-0c63-4c98-2ce9-08e1e1796f50"
+          "detection_log": [],
```
This should not be done silently, we should have the `extra-words` in the detection log and add another entry like `extra-words-permitted-in-rule` to make sure this is done explicitly and logged correctly.

alok1304 replied:
yes i will do that

AyanSinhaMahapatra commented on src/licensedcode/match.py:
```diff
@@ -1071,6 +1077,53 @@ def merge_matches(matches, max_dist=None, trace=TRACE_MERGE):
 # early from the loops: trying to check containment on wildly separated matches
 # does not make sense
 
+def is_extra_words_position_valid(match):
```
We need some unit tests for this function

AyanSinhaMahapatra commented on src/licensedcode/match.py:
```diff
@@ -598,6 +598,12 @@ def score(self):
+
+        # Check whether extra words in the matched text appear in allowed positions,
+        # and do not exceed the maximum allowed word count at those positions.
+        if is_extra_words_position_valid(self):
```
I'm not sure if we want to change the score directly at this stage. This should be done in detection post-processing ideally (so we can log and don't do this silently), from here: https://github.com/aboutcode-org/scancode-toolkit/blob/38cf1ad31e133df11386d28cf6d89a33f16f423f/src/licensedcode/detection.py#L1576

This is where you analyze the matches for detection issues, and fix accordingly.

alok1304 replied:
I will see

AyanSinhaMahapatra commented on tests/licensedcode/data/plugin_license/extra-words/scan-extra-words-2-aho-license.expected.json:
```diff
@@ -16,7 +14,7 @@
           "start_line": 4,
           "end_line": 27,
           "matcher": "2-aho",
```
This seems to work for `2-aho` but I don't think you've changed the `3-seq` test, did that not work correctly as expected?

The score sees to be unchanged there: https://github.com/alok1304/scancode-toolkit/blob/extra_words/tests/licensedcode/data/plugin_license/extra-words/scan-extra-words-3-seq-license.expected.json#L49 
Could you debug what happened?

alok1304 replied:
In `3-seq` it matches with other rules which have a different position of `extra-words`. I will do that for this, I will mark another position where `extra-words` may be present.

AyanSinhaMahapatra commented on src/licensedcode/data/rules/bsd-new_158.RULE:
```diff
@@ -14,7 +14,7 @@ Redistributions in binary form must reproduce the above copyright
 documentation and/or other materials provided with the distribution.
 
-Neither the name of nor the names of its
+Neither the name of [[3]] nor the names of its
```
```suggestion
Neither the name of [[6]] nor the names of its
```
Let's increase this a bit just in case

alok1304 replied:
okay

AyanSinhaMahapatra reviewed (changes requested):
@alok1304 Thanks++ this is a great start, see comments for changes requested.
See also the failure at https://dev.azure.com/nexB/scancode-toolkit/_build/results?buildId=16448&view=logs&j=1e361573-c314-56d5-4489-17c77c3ccfe6&t=b85906fe-175b-5369-bd04-3d39d845e187&l=14890 where we check that our rules are detected correctly, and this is failing because the changed rule is being detected with `ectra-words` and we are not handling the `3-seq` case correctly yet.

alok1304 committed:
add test for correct position of `extra-words` and enhance `detection_log`

Add test for is correct position of `extra-words` according to `extra-phrases` that is present in rules.

if we find `extra-words` are in the right place then we set score to `100`.
And also show in `detection_log` why we increasing the score to keep track of this.

Signed-off-by: Alok Kumar <alokkumarjipura9973@gmail.com>

alok1304 commented:
i will close this pr starting with new pr

alok1304 committed:
Merge branch 'develop' into extra_words

alok1304 committed:
Update tallies_key_files-details.expected.json-lines
